tsl/src/nodes/skip_scan/exec.c

```diff
@@ -0,0 +1,597 @@
+		for (int i = 0; i < num_runtime_keys; i++)
+		{
+			if (runtime_keys[i].scan_key <= skip_key)
+				runtime_keys[i].scan_key -= 1;
+		}
+	}
+}
+
+/****************************
+ * small accessor functions *
+ ****************************/
+
+static inline bool
+skip_scan_is_searching_for_first_val(SkipScanState *state)
+{
+	return state->stage == SkipScanSearchingForFirst;
+}
+
+static inline bool
+skip_scan_state_found_null(SkipScanState *state)
+{
+	return (state->stage & SkipScanFoundNull) != 0;
+}
+
+static inline bool
+skip_scan_state_found_val(SkipScanState *state)
+{
+	return (state->stage & SkipScanFoundVal) != 0;
+}
+
+static inline bool
+skip_scan_state_found_everything(SkipScanState *state)
+{
+	return (state->stage & SkipScanFoundNullAndVal) == SkipScanFoundNullAndVal;
+}
+
+static inline bool
+skip_scan_state_is_searching_for_null(SkipScanState *state)
+{
+	return (state->stage & SkipScanSearchingForNull) == SkipScanSearchingForNull;
+}
+
+static inline bool
+skip_scan_state_is_searching_for_val(SkipScanState *state)
+{
+	return (state->stage & SkipScanSearchingForVal) == SkipScanSearchingForVal;
+}
+
+static inline bool
+skip_scan_is_finished(SkipScanState *state)
+{
+	/*
+	 * Once the underlying Index(Only)Scan runs out of tuples,
+	 * we're not going to find anything more if
+	 *  1. We're searching for first, and found nothing:
+	 *     the regular qual must exclude everything
+	 *  2. We're searching for a NULL, but have not found one:
+	 *     we must have already found a non-NULL val, and be searching for a final NULL
+	 *  3. We're searching for a non-NULL, but have not found one:
+	 *     we must have already found a NULL val, and be searching for non-NULL ones
+	 */
+	return (state->stage & SkipScanFoundNullAndVal) == 0 ||
+		   (skip_scan_state_is_searching_for_val(state) && !skip_scan_state_found_val(state)) ||
+		   (skip_scan_state_is_searching_for_null(state) && !skip_scan_state_found_null(state));
+}
+
+static inline IndexScanDesc
+skip_scan_state_get_scandesc(SkipScanState *state)
+{
+	return *state->scan_desc;
+}
+
+static inline ScanKey
+skip_scan_state_get_scankeys(SkipScanState *state)
+{
+	return *state->scan_keys;
+}
+
+static inline ScanKey
+skip_scan_state_get_skipkey(SkipScanState *state)
+{
+	Assert(!state->skip_qual_removed);
+	Assert(*state->num_scan_keys > 0);
+	return &skip_scan_state_get_scankeys(state)[state->skip_qual_offset];
+}
+
+/*******************************
+ * Primary Execution Functions *
+ *******************************/
+
+static TupleTableSlot *
+skip_scan_exec(CustomScanState *node)
+{
+	SkipScanState *state = (SkipScanState *) node;
+	TupleTableSlot *result;
+
+	if (skip_scan_is_searching_for_first_val(state))
+	{
+		Assert(skip_scan_state_get_scandesc(state) == NULL);
+		/* first time through we ignore the inital scan keys, which are used to
+		 * skip previously seen values, we'll change back the number of scan keys
+		 * the first time through update_skip_key
+		 */
```

Comment:
```suggestion
		/* first time through we ignore the skip scan keys, which are used to
		 * skip previously seen values, we'll change back the number of scan keys
		 * the first time through update_skip_key
		 */
```
This is no longer accurate as the skip scan key does not need to be the first key